### PR TITLE
Add API to configure the fallback fonts provider when all default methods return null 

### PR DESCRIPTION
**Pull request by eymar**
New API:

```kotlin
class TypefaceFontProviderWithFallback
class FontMgrWithFallback
```

Usage:

```kotlin
val fontCollection = FontCollection()

val notoEmojiTypeface = Typeface.makeFromData(Data.makeFromBytes(notoEmojisBytes))
val fallbackProvider = TypefaceFontProviderWithFallback().apply {
    registerTypeface(notoEmojiTypeface)
}

fontCollection.setDefaultFontManager(FontMgrWithFallback(fallbackProvider))
```

Screenshot 2024-06-07 at 14 48 28

### CODE REVIEW

**Comment by eymar** [comment]
Without fallback or with only 1 font registered in fallback provider: 

Screenshot 2024-06-07 at 15 40 23
Screenshot 2024-06-07 at 15 40 39
Screenshot 2024-06-07 at 15 40 50

File: skiko/src/jvmMain/cpp/common/paragraph/TypefaceFontProvider.cc [hunk 1]
```diff
@@ -12,7 +13,7 @@ extern "C" JNIEXPORT jlong JNICALL Java_org_jetbrains_skia_paragraph_TypefaceFon
     return reinterpret_cast<jlong>(instance);
 }
 
-extern "C" JNIEXPORT void JNICALL Java_org_jetbrains_skia_paragraph_TypefaceFontProviderKt__1nRegisterTypeface
+extern "C" JNIEXPORT jint JNICALL Java_org_jetbrains_skia_paragraph_TypefaceFontProviderKt__1nRegisterTypeface
```
Comment (MatkovIvan) [inline skiko/src/jvmMain/cpp/common/paragraph/TypefaceFontProvider.cc:16]: There is return for this int

File: skiko/src/commonMain/kotlin/org/jetbrains/skia/paragraph/FontCollection.kt [hunk 2]
```diff
@@ -106,7 +106,6 @@ class FontCollection internal constructor(ptr: NativePointer) : RefCnt(ptr) {
             reachabilityBarrier(this)
         }
     }
-
```
Comment (MatkovIvan) [inline skiko/src/commonMain/kotlin/org/jetbrains/skia/paragraph/FontCollection.kt:109]: Please revert such formatting (no changes in this file)

File: skiko/src/jvmMain/cpp/common/paragraph/TypefaceFontProvider.cc (hunk 1 above)
Comment (eymar) [reply skiko/src/jvmMain/cpp/common/paragraph/TypefaceFontProvider.cc:16]: no compilation error despite this. 
thanks! fixed

File: samples/SkiaMultiplatformSample/src/wasmJsMain/kotlin/org/jetbrains/skiko/sample/Utils.kt [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import org.w3c.xhr.XMLHttpRequest
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+import kotlin.wasm.unsafe.UnsafeWasmMemoryApi
+import kotlin.wasm.unsafe.withScopedMemoryAllocator
+
+suspend fun loadRes(url: String): ArrayBuffer {
+    return suspendCoroutine { continuation ->
+        val req = XMLHttpRequest()
```
Comment (Schahen) [inline samples/SkiaMultiplatformSample/src/wasmJsMain/kotlin/org/jetbrains/skiko/sample/Utils.kt:14]: nothing blocking, just out of pure interest - why not fetch?  Worser external declarations in js/wasm?

File: skiko/src/commonTest/kotlin/org/jetbrains/skia/FontFallbackTest.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            val df = it.defaultFallback(0x6C34, FontStyle.NORMAL, null)
+            val glyphs = df!!.getStringGlyphs("水")
+
+            if (kotlinBackend.isWeb()) {
+                assertEquals("Noto Sans SC", df.familyName)
+                assertContentEquals(shortArrayOf(16391), glyphs)
+            } else {
+                // the actual font can be different on different systems,
+                // so just making sure it's not null
+                assertNotEquals(null, df?.familyName)
```
Comment (Schahen) [inline skiko/src/commonTest/kotlin/org/jetbrains/skia/FontFallbackTest.kt:35]: looks like safe-call is redundant here

File: samples/SkiaMultiplatformSample/src/wasmJsMain/kotlin/org/jetbrains/skiko/sample/Utils.kt (hunk 3 above)
Comment (eymar) [reply samples/SkiaMultiplatformSample/src/wasmJsMain/kotlin/org/jetbrains/skiko/sample/Utils.kt:14]: copy pasted from another place without thinking. Will fix :)

File: skiko/src/commonTest/kotlin/org/jetbrains/skia/FontFallbackTest.kt [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            }
+        }
+    }
+
+    @Test
+    fun testChinese_noFallback() = runTest {
```
[the remark this answers is not in the document]
Comment (eymar) [reply skiko/src/commonTest/kotlin/org/jetbrains/skia/FontFallbackTest.kt:41]: it's possible. we have a common test source for js and Wasm

File: skiko/src/commonTest/kotlin/org/jetbrains/skia/FontFallbackTest.kt (hunk 4 above)
Comment (eymar) [reply skiko/src/commonTest/kotlin/org/jetbrains/skia/FontFallbackTest.kt:35]: I agree with you. But the compiler doesn't like it:

> Type inference failed. The value of the type parameter T should be mentioned in input types (argument types, receiver type or expected type). Try to specify it explicitly.


If I add the type explicitly like this: `assertNotEquals<String?>(null, df.familyName)` 
it says that I can remove explicit type arguments. When I do, it shows the error as before.

Looks like a bug.

File: samples/SkiaMultiplatformSample/src/wasmJsMain/kotlin/org/jetbrains/skiko/sample/Utils.kt (hunk 3 above)
Comment (eymar) [reply samples/SkiaMultiplatformSample/src/wasmJsMain/kotlin/org/jetbrains/skiko/sample/Utils.kt]: fixed

File: skiko/src/commonTest/kotlin/org/jetbrains/skia/FontFallbackTest.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (eymar) [reply skiko/src/commonTest/kotlin/org/jetbrains/skia/FontFallbackTest.kt]: fixed

File: skiko/src/commonTest/kotlin/org/jetbrains/skia/FontFallbackTest.kt (hunk 4 above)
Comment (eymar) [reply skiko/src/commonTest/kotlin/org/jetbrains/skia/FontFallbackTest.kt]: changed to explicit type and removed a safe-call

**Review by Schahen — APPROVED** [review]

File: skiko/src/commonMain/kotlin/org/jetbrains/skia/paragraph/TypefaceFontProvider.kt [hunk 6]
```diff
@@ -12,22 +12,40 @@ import org.jetbrains.skia.impl.NativePointer
...
+) : FontMgr(ptr) {
+
+    constructor(): this(TypefaceFontProvider_nMake(), false)
+
     companion object {
         init {
             staticLoad()
         }
+
+        fun createAsFallbackProvider(): TypefaceFontProvider {
```
Comment (igordmn) [inline skiko/src/commonMain/kotlin/org/jetbrains/skia/paragraph/TypefaceFontProvider.kt:27]: It looks like there is no reason for user to not always create `TypefaceFontProvider.createAsFallbackProvider`. The API and behaviour is the same, the only difference is that the fallback one can be passed to another function.

We can just pass `TypefaceFontProvider_nMakeAsFallbackProvider` in the constructor, and remove this function (and other unused functions/type checks after that).

File: skiko/src/commonMain/kotlin/org/jetbrains/skia/FontMgr.kt [hunk 7]
```diff
@@ -10,6 +11,13 @@ open class FontMgr : RefCnt {
         }
 
         val default = FontMgr(_nDefault(), false)
+
+        fun defaultWithFallbackFontProvider(fallback: TypefaceFontProvider): FontMgr {
```
Comment (igordmn) [inline skiko/src/commonMain/kotlin/org/jetbrains/skia/FontMgr.kt:15]: The functionality of adding fallback seems doesn't relate to the default FontMgr, and the API looks unnecessary restricted from an external user perspective (allowing fallbacks only for the default manager).

I would make a separate extended class instead:
```
class FontMgrWithFallback: FontMgr() {
    constructor(fallback: TypefaceFontProvider): this(_nDefaultWithFallbackFontProvider(fallback._ptr))
}
```

This also mimics the C++ API, which is a good thing for keeping Skiko API as close as possible to the original.

User can use the API this way then:
```
val fallbackProvider = TypefaceFontProvider().apply {
    registerTypeface(notoEmojiTypeface)
}

fontCollection.setDefaultFontManager(FontMgrWithFallback(FontMgr.default, fallbackProvider))
```

**Review by igordmn — COMMENTED** [review]
Thanks!

Looks good, but the API can be improved.

File: skiko/src/commonMain/kotlin/org/jetbrains/skia/FontMgr.kt (hunk 7 above)
Comment (eymar) [reply skiko/src/commonMain/kotlin/org/jetbrains/skia/FontMgr.kt:15]: The usage is a bit different though:

the default `TypefaceFontProvider` won't provide the fallbacks.
But `TypefaceFontProvider.createAsFallbackProvider()` will provide them.

Comment (eymar) [reply skiko/src/commonMain/kotlin/org/jetbrains/skia/FontMgr.kt]: - add a new `TypefaceFontProviderWithFallback` it extends the existing `TypefaceFontProvider`
- `TypefaceFontProvider` is now an open class
- added a new`class FontMgrWithFallback(fallbackProvider: TypefaceFontProviderWithFallback)`, which relies on skia's FontMgr.default under the hood + our new fallback provider

File: skiko/src/commonMain/kotlin/org/jetbrains/skia/paragraph/TypefaceFontProviderWithFallback.kt [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import org.jetbrains.skia.ExternalSymbolName
+import org.jetbrains.skia.FontMgr
+import org.jetbrains.skia.ModuleImport
+import org.jetbrains.skia.Typeface
+import org.jetbrains.skia.impl.*
+import org.jetbrains.skia.impl.getPtr
+import org.jetbrains.skia.impl.interopScope
+import org.jetbrains.skia.impl.reachabilityBarrier
+
+class TypefaceFontProviderWithFallback private constructor(
```
[the remark this answers is not in the document]
Comment (eymar) [reply skiko/src/commonMain/kotlin/org/jetbrains/skia/paragraph/TypefaceFontProviderWithFallback.kt:12]: Thanks! fixed now.

**Review by igordmn — APPROVED** [review]

**Pull request merged**